### PR TITLE
Load parent report action from withOnyx in report action HOC

### PR DESCRIPTION
**Pull request by tgolen**
@luacmartins I wanted you to review this to ensure I didn't end up changing any of the logic and it looked like you had the most context into the current code.

### Fixed Issues
Part of https://github.com/Expensify/App/issues/27262

### Tests
1. Run `npm i` and also a `pod install` from the `ios` directory
1. Open a chat with Person A and Person B
2. Have Person B open a thread from one of Person A's comments and add a message to the thread
3. From inside the thread, have Person B flag the original comment from Person A (the topmost comment)
4. Verify the Flag Comment window opens with options (if this PR were not working, this would display as a `not-found` page instead)

- [ ] Verify that no errors appear in the JS console

### Offline tests
Same as above

### QA Steps
1. Open a chat with Person A and Person B
2. Have Person B open a thread from one of Person A's comments and add a message to the thread
3. From inside the thread, have Person B flag the original comment from Person A (the topmost comment)
4. Verify the Flag Comment window opens with options (if this PR were not working, this would display as a `not-found` page instead)

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome UNABLE TO TEST DUE TO BROKEN EMULATOR
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

image

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

image

</details>

<details>
<summary>iOS: mWeb Safari</summary>

image

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

image

</details>

<details>
<summary>MacOS: Desktop</summary>

image

</details>

### CODE REVIEW

**Comment by tgolen** [comment]
Got stuck with the TS errors. There is a draft PR for Onyx which will probably be required: https://github.com/Expensify/react-native-onyx/pull/445

**Comment by tgolen** [comment]
Now that https://github.com/Expensify/react-native-onyx/pull/446 is merged, I need to update Onyx to 1.0.126 and finish this up.

File: src/pages/FlagCommentPage.js [hunk 1]
```diff
@@ -200,10 +199,4 @@ FlagCommentPage.displayName = 'FlagCommentPage';
 export default compose(
     withLocalize,
     withReportAndReportActionOrNotFound,
-    withOnyx({
```
Comment (tgolen) [inline src/pages/FlagCommentPage.js:203]: This can be removed because the `parentReportAction` comes from `withReportAndReportActionOrNotFound` now.

**Review by luacmartins — COMMENTED** [review]
Changes LGTM. Agree with @amyevans that we should merge main to get rid of the Onyx diff.

**Comment by tgolen** [comment]
All right, I updated with `main`. Hopefully, that gets everything fixed. If not, I'll need to come back to this in the afternoon to wrap up any final issues.

**Review by luacmartins — DISMISSED** [review]
LGTM

File: src/pages/FlagCommentPage.js [hunk 2]
```diff
@@ -197,13 +194,4 @@ FlagCommentPage.propTypes = propTypes;
...
-    withLocalize,
-    withReportAndReportActionOrNotFound,
-    withOnyx({
-        parentReportActions: {
-            key: ({report}) => `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${report.parentReportID || report.reportID}`,
-            canEvict: false,
-        },
-    }),
-)(FlagCommentPage);
+export default compose(withLocalize, withReportAndReportActionOrNotFound)(FlagCommentPage);
```
[the remark this answers is not in the document]
Comment (tgolen) [reply src/pages/FlagCommentPage.js:197]: Thanks, I don't want to add too many extra things to these PRs. They get difficult to review and test then.

**Comment by tgolen** [comment]
@aimane-chnaif I tried to look into that bug and I couldn't reproduce it on web or on iOS native. Not only couldn't I see the not-found page happening, but I also added a console log inside the same `if` statement that renders the `not-found` page and I never saw the log being triggered. Is there any debugging you can help me with to see why you're seeing that?

**Comment by tgolen** [comment]
To be clear, this is where I was adding the console log: https://github.com/Expensify/App/blob/9cf1ed840158a6e5e536a38b791be6cf39ae7416/src/pages/home/report/withReportAndReportActionOrNotFound.tsx#L85

**Comment by tgolen** [comment]
@aimane-chnaif Can you please link me to something that tracks the revert of the Onyx bump? I am not really sure how best to follow that to know when this can be taken off HOLD.

**Comment by luacmartins** [comment]
@tgolen it seems like the [PR bumping Onyx got merged/deployed](https://github.com/Expensify/App/pull/34801). I think we can remove the hold here

**Comment by tgolen** [comment]
OK, I'll update to `2.0.1` and take this off HOLD later today. Thank you for thinking of me!

**Comment by tgolen** [comment]
This is updated and off HOLD so it is ready for review/test/merge again. There were no code changes other than the bump of the onyx version, which was applied by merging `main` into this branch.

**Comment by tgolen** [comment]
Ah phooey, it looks like `WithOnyxInstanceState` is no longer exported and I've got to update Onyx to export that type 😦

**Comment by tgolen** [comment]
I'm putting this back on hold while I do https://github.com/Expensify/react-native-onyx/pull/463

**Comment by tgolen** [comment]
All right, this is off HOLD again and ready for review/testing @aimane-chnaif. Thanks for being patient on this!

**Review by luacmartins — DISMISSED** [review]
LGTM!

File: src/pages/home/report/withReportAndReportActionOrNotFound.tsx [hunk 3]
```diff
@@ -114,6 +116,17 @@ export default function <TProps extends ComponentProps, TRef>(WrappedComponent:
                 key: ({route}) => `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${route.params.reportID}`,
                 canEvict: false,
             },
+            parentReportAction: {
+                key: (props) => `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${props?.report?.parentReportID ?? 0}`,
+                selector: (parentReportActions: OnyxEntry<OnyxTypes.ReportActions>, props: WithOnyxInstanceState<OnyxProps>): OnyxEntry<OnyxTypes.ReportAction> => {
+                    const parentReportActionID = props?.report?.parentReportActionID;
```
[the remark this answers is not in the document]
Comment (tgolen) [reply src/pages/home/report/withReportAndReportActionOrNotFound.tsx:126]: I've updated the branch with this, but it's throwing a TS error that I can't figure out. @aimane-chnaif do you have any idea how to fix it? If not, I'll ask in Slack.

File: src/pages/home/report/withReportAndReportActionOrNotFound.tsx [hunk 4]
```diff
@@ -94,24 +98,37 @@ export default function <TProps extends ComponentProps, TRef>(WrappedComponent:
     WithReportOrNotFound.displayName = `withReportOrNotFound(${getComponentDisplayName(WrappedComponent)})`;
 
     return compose(
-        withOnyx<TProps & RefAttributes<TRef>, OnyxProps>({
+        withOnyx<TProps & RefAttributes<TRef>, OnyxPropsA>({
```
[the remark this answers is not in the document]
Comment (tgolen) [reply src/pages/home/report/withReportAndReportActionOrNotFound.tsx:101]: I believe so. I am going to ping @blazejkustra and @fabioh8010 to help out with this one.

**Review by luacmartins — COMMENTED** [review]
We have a few tests failing.

File: src/pages/home/report/withReportAndReportActionOrNotFound.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (tgolen) [reply src/pages/home/report/withReportAndReportActionOrNotFound.tsx:101]: Posted to slack [here](https://expensify.slack.com/archives/C0593T50UHG/p1707150835939009) for discussion.

**Comment by tgolen** [comment]
Yeah, there is a TS error that we're trying to get figured out. I'll hopefully have this wrapped up today.

**Comment by tgolen** [comment]
@aimane-chnaif after struggling with the TS stuff, I decided to go back to using only a single `withOnyx` wrapper. I added some logic to improve when the not-found page is shown so that it hopefully won't be an issue like you saw.

File: src/pages/home/report/withReportAndReportActionOrNotFound.tsx [hunk 5]
```diff
@@ -78,7 +80,9 @@ export default function <TProps extends ComponentProps, TRef>(WrappedComponent:
         }
 
         // Perform the access/not found checks
-        if (shouldHideReport || isEmptyObject(reportAction)) {
+        // Be sure to avoid showing the not-found page while the parent report actions are still being read from Onyx. The parentReportAction will be undefined while it's being read from Onyx
+        // and then reportAction it will either be a valid parentReportAction or an empty object. In the case of an empty object, then it's OK to show the not-found page.
```
Comment (luacmartins) [inline src/pages/home/report/withReportAndReportActionOrNotFound.tsx:84]:
```suggestion
        // and then reportAction will either be a valid parentReportAction or an empty object. In the case of an empty object, then it's OK to show the not-found page.
```

**Review by luacmartins — COMMENTED** [review]
Code LGTM, just one minor comment change.

**Review by luacmartins — DISMISSED** [review]

**Review by aimane-chnaif — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/96077027/e3026f29-6963-4e5c-9212-0f1f04f48cf4



</details>



<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/96077027/3c414456-a8e1-4479-853c-399100593b70



</details>

<details>
<summary>iOS: Native</summary>


https://github.com/Expensify/App/assets/96077027/19c4c92b-9cb9-414b-af27-0357fb500ac2



</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/Expensify/App/assets/96077027/62a9cf35-509e-4c22-9745-658854391638



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



https://github.com/Expensify/App/assets/96077027/a1eb4249-d7a6-4c4b-a2e2-64b4d956e659


</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/Expensify/App/assets/96077027/a152c7a8-bacc-43c8-a100-64ca9498171c



</details>

**Review by luacmartins — COMMENTED** [review]
Seems like we have an edge case

**Comment by tgolen** [comment]
I think that edgecase is fine. Do either of you feel strongly about improving that? I'll fix the conflicts this morning.

**Comment by tgolen** [comment]
Practically thinking through it, when is a real user ever going to enter that flow? And if they do, what are the consequences when they do? I think it's highly unlikely anyone would ever encounter it in the real world, and the worst that can happen when they do is they see a not-found page while the report loads. IMO it is too low value to worry about. I would also explain that to anyone who reports it as a regression.

**Review by luacmartins — APPROVED** [review]

**Comment by tgolen** [comment]
Thanks for hearing me out.

I fixed the conflicts. It was caused because both files were migrated to TS. That probably means it's worth having another last look at the code before shipping it.

**Review by aimane-chnaif — COMMENTED** [review]
:shipit: 


https://github.com/Expensify/App/assets/96077027/d16216d2-6455-41ac-bf2a-979d25787e1c

**Comment by luacmartins** [comment]
Thanks! Merging!

**Pull request merged**